plugins/action/win_pssession_configuration.py

```diff
@@ -0,0 +1,110 @@
+# Copyright: (c) 2018, Ansible Project
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+import ntpath
+import time
+
+from ansible.errors import AnsibleError
+from ansible.plugins.action import ActionBase
+from ansible.utils.vars import merge_hash
+from ansible.utils.display import Display
+display = Display()
+
+
+class ActionModule(ActionBase):
+    _default_async_timeout = 300
+    _default_async_poll = 1
+
+    def __init__(self, *args, **kwargs):
+        super(ActionModule, self).__init__(*args, **kwargs)
+
+    def run(self, tmp=None, task_vars=None):
+        self._supports_check_mode = True
+        self._supports_async = True
+        check_mode = self._play_context.check_mode
+        async_timeout = self._task.async_val
```

Comment:
Yeah, sort of. A lot of manipulation of the "real" `async` and `poll` values has to be done, so this is to save the value passed in.

It does also get used later, as the real `async` value for the module invocation. In that way, I think letting the user use the standard `async` keyword is the most intuitive. 

With a separate module option, it would be a bit confusing if a user used the keyword, or used both the module option and keyword.

Since I need to do some manual setting of the attributes on `self._task` anyway I can't really avoid touching those fields.

If it's not a dealbreaker for you I'd prefer to keep it as is and flow through user-specified `async` and `poll` keyword values.

Reply:
I would strong urge you against it. If we ever start supporting running action plugins asynchronously and not just the module part then this module will never work as you've co-opted the `async` and `poll` task directives for use inside the action plugin and not towards it. While you probably can't avoid touching `self._task` entirely it should be a end goal to avoid using it as much as you can.

It may even be more beneficial for you to run the module in the action plugin without restarting the WinRM service, execute the service restart and then wait until the service is back up and running similar to how `win_reboot` works instead of trying to run things entirely through async but I don't think that is necessary right now.

Reply:
I can't avoid restarting the WinRM service because the cmdlets that modify the session configurations don't properly support it. They have a switch for not restarting the service but it doesn't actually work, and the service is restarted anyway (I originally approached the problem that way as it would have been much cleaner, but unfortunately it's not possible).

Is there a proposal/PR for making actions async? It might be possible to support both depending on how its implemented.

Anyway I'll look into changing these values into a module option instead for now. They'll have to be used in the action plugin in basically the same way as now, but if it makes for better future proofing I'll go that route.

Reply:
> I can't avoid restarting the WinRM service because the cmdlets that modify the session configurations don't properly support it

The module could still set things up and return a command that needs to be run for `Register-PSSessionConfiguration` and the "service restart" step is just running that particular command.

> Is there a proposal/PR for making actions async? It might be possible to support both depending on how its implemented.

No proposal but it should be something to think about before trying to reuse an existing task directive(s) for an internal purpose.

Reply:
> The module could still set things up and return a command that needs to be run for `Register-PSSessionConfiguration` and the "service restart" step is just running that particular command.

Hm that's interesting. How would you recommend running the command itself? Invoking `raw` / `win_shell` basically and then catching the exception and doing `wait_for_connection` maybe?

Sidestepping the need for async at all does sound attractive. 

I'll look into doing that vs. switching away from the keywords toward module options; might want to do that in a future revision if I end up running into unexpected issues, but I like the idea a lot.

Reply:
I started looking into that idea but realized it would be a little more involved than I initially thought, so I've gone the other way of adding async control options to the module.

To help alleviate incompatibility with future async enhancements, the module now reports that it doesn't support async, so that should solve any weirdness in trying to use the async keywords along with or instead of the new module options (it will fail).

The documentation has been updated to reflect this, and there are a few new tests to validate the async options and that the async keywords fail as expected.